### ISSUE TITLE
Nullness Checker allows for dereference of a @Nullable local variable

### ISSUE THREAD

**Issue by PawelLipski** [typetools/checker-framework#4079]
A repro is included in https://github.com/VirtusLab/git-machete-intellij-plugin/pull/672/files. I can provide a more stripped-down repro if needed.

In a nutshell, I have a piece like that:

```java
      // org.eclipse.jgit.revwalk.RevWalk walk;
      @Nullable RevCommit mergeBase = walk.next();
      LOG.debug(() -> "Detected merge base for " + c1.getHash().getHashString() + " and " + c2.getHash().getHashString()
          + " is " + mergeBase.getId().getName());
```

Clearly, calling `mergeBase.getId()` should not be allowed... yet the compilation passes without an error.

`RevWalk#next` is declared as:

```java
	public RevCommit next() throws MissingObjectException,
			IncorrectObjectTypeException, IOException {
		return pending.next();
	}
```

I've also added `-AuseConservativeDefaultsForUncheckedCode=+bytecode`, to no avail :/

**Comment by mernst** [comment]
This behavior looks right to me.

 * `next()` is declared to return a `@NonNull RevCommit`.
 * `walk.next()` has type `@NonNull RevCommit`.
 * `mergeBase` is declared as type `@Nullable RevCommit`.  However, it is assigned a non-null value, so the Nullness Checker treats it as `@NonNull RevCommit` from that assignment until it might be re-assigned.

You can see examples that are similar to yours in the [type refinement section of the manual](https://checkerframework.org/manual/#type-refinement-examples).

If `RevWalk.next()` might return null, can you declare its return type as `@Nullable`?
Or if you cannot do that, can you write a stub file to set its type?
If you are trying to use `-AuseConservativeDefaultsForUncheckedCode=+bytecode`, is `RevWalk` read from bytecode?

**Comment by PawelLipski** [comment]
> If you are trying to use -AuseConservativeDefaultsForUncheckedCode=+bytecode, is RevWalk read from bytecode?

Yes, that's exactly the case, it's coming from JGit... and unfortunately, its javadoc tells it can return null:

![image](https://user-images.githubusercontent.com/3383210/103202593-1fa39880-48f3-11eb-8043-c5d0770fe989.png)

but JGit methods aren't marked with any nullability annotations.

It seems that NullnessChecker unsoundly (?) assumes that the method returns a non-null value, despite `-AuseConservativeDefaultsForUncheckedCode=+bytecode`... lemme change the title of this issue...

**Comment by PawelLipski** [comment]
Actually... TBH I've never noticed that turning on `-AuseConservativeDefaultsForUncheckedCode` changes anything in the compilation (no extra errors show up), which is suspicious 🤔

**Comment by PawelLipski** [comment]
@mernst hey, any extra details needed here? I can provide a stripped down repro in a separate repo if needed
